sdk/generic-sdk-core/generic-core/src/main/java/com/generic/core/http/client/DefaultHttpClient.java

```diff
@@ -194,33 +201,209 @@ private void sendBody(HttpRequest httpRequest, Object progressReporter, HttpURLC
+        RetrySSEResult retrySSEResult;
+        try (BufferedReader reader
+                 = new BufferedReader(new InputStreamReader(connection.getInputStream()))) {
+            retrySSEResult = processBuffer(reader, listener);
+            if (retrySSEResult != null) {
+                retryExceptionForSSE(retrySSEResult, listener, httpRequest);
+            }
+        } catch (IOException e) {
+            throw LOGGER.logThrowableAsError(new UncheckedIOException(e));
+        }
+    }
 
-            try (InputStream errorStream = connection.getErrorStream();
-                InputStream inputStream = (errorStream == null) ? connection.getInputStream() : errorStream) {
-                byte[] buffer = new byte[8192];
-                int length;
-                while ((length = inputStream.read(buffer)) != -1) {
-                    outputStream.write(buffer, 0, length);
+    private boolean isTextEventStream(Headers responseHeaders) {
+        return responseHeaders.get(HeaderName.CONTENT_TYPE) != null &&
+        responseHeaders.get(HeaderName.CONTENT_TYPE).getValue().equals(ContentType.TEXT_EVENT_STREAM);
+    }
+
+    /**
+     * Processes the sse buffer and dispatches the event
+     *
+     * @param reader   The BufferedReader object
+     * @param listener The listener object attached with the httpRequest
+     */
+    private RetrySSEResult processBuffer(BufferedReader reader, ServerSentEventListener listener) {
+        StringBuilder collectedData = new StringBuilder();
+        ServerSentEvent event = null;
+        try {
+            String line;
+            while ((line = reader.readLine()) != null) {
+                collectedData.append(line).append("\n");
+                if (isEndOfBlock(collectedData)) {
+                    event = processLines(collectedData.toString().split("\n"));
+                    if (!Objects.equals(event.getEvent(), DEFAULT_EVENT) || event.getData() != null) {
+                        listener.onEvent(event);
+                    }
+                    collectedData.setLength(0); // clear the collected data
                 }
             }
-
-            return new DefaultHttpClientResponse(httpRequest, responseCode, responseHeaders,
-                BinaryData.fromByteBuffer(outputStream.toByteBuffer()));
+            listener.onClose();
         } catch (IOException e) {
-            throw LOGGER.logThrowableAsError(new RuntimeException(e));
-        } finally {
-            connection.disconnect();
+            return new RetrySSEResult(e, event != null ? event.getId() : -1, event != null ? event.getRetryAfter() : null);
+        }
+        return null;
+    }
+
+    private boolean isEndOfBlock(StringBuilder sb) {
+        // blocks of data are separated by double newlines
+        // add more end of blocks here if needed
+        return sb.indexOf("\n\n") >= 0;
```

Comment:
How does BufferedReader.readLine() work when there are multiple line returns in a row? And are we better off tracking how many empty lines in a row we receive instead of this?